README.md
What is this?
=============

Write is a word processor activity for the Sugar desktop.

Write embeds the AbiWord word processor, and can be used to write and edit text documents.

How to use?
===========

Write is part of the Sugar desktop and is always included.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/), and;
* [How to use Write](https://help.sugarlabs.org/en/write.html).

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/en/my_settings.html), [Software Update](https://help.sugarlabs.org/en/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `Write`, then download.

How to integrate?
=================

Write depends on AbiWord, Python, [Sugar Toolkit for GTK+ 3](https://github.com/sugarlabs/sugar-toolkit-gtk3), GStreamer 1, GTK+ 3, Telepathy, and gst-plugins-espeak.

Write is started by [Sugar](https://github.com/sugarlabs/sugar).

Write is [packaged by Fedora](https://src.fedoraproject.org/rpms/sugar-write).  On Fedora systems;

```
dnf install sugar-write
```

Write is packaged by Debian and Ubuntu distributions.  On Debian and Ubuntu systems;

```
apt install sugar-write-activity
```


NEWS
101

* New superscript and subscript buttons (Andy Tockman, Swarup N),
* Remove empty locale files (James Cameron),
* Port to Python 3 (Rahul Bothra, James Cameron),
* Port to sugar3.speech (James Cameron),

100

* Fix create table function (Swarup N),
* Add merge and split table cells (Ezequiel Pereira López, Swarup N),
* Update POT file (James Cameron),
* New translations (Chris Leonard et al).

99.1

* Add paste unformatted option (Ezequiel Pereira),
* Demote errors to debug (James Cameron),
* Port to TelepathyGLib (Rahul Bothra),
* Add screenshots and update metadata (Vipul Gupta),
* Update POT file (James Cameron).

99

* Use menu item instead of checkbox for image palette (Sam Parkinson),
* Improve grid creator UI (Sam Parkinson),
* Port from GConf (James Cameron),
* Fix use of Gst.Message.structure (James Cameron),
* Update POT file (James Cameron).

98.4

* New translations (Chris Leonard et al)

98.3

* Fix initial state of Undo and Redo buttons (James Cameron)

98.2

* Fix for fail to start with Abiword 2.9 on Fedora 18

98.1

* Update POT file
* Fix Python GI API version warnings

98

* Update POT and refresh from Pootle, trim headers (Chris Leonard)
* Yoruba (iamutkarshtiwari)
* Added max_participants = 6 in activity.info (AbrahmAB)
* Fix crash on startup due to missing module initialization. (Icarito)
* fix repository url (James Cameron)

66

* Revert export button

65

* #1145 Do not propagate current paragraph format while moving cursor

64

* Utilize new toolbars
* #1121 Write canvas redraw glitch when locking open a secondary toolbar
* #1120 Fix MANIFEST file
* #1122 New Write primary toolbar icon design does not all fit in the toolbar

58

* Fix 6021: Write crash on buddy left
* Add license field to activity.info

57

* Translation release

56

* Translation release

55

* Translation release

54

* Fix 4871: the write save as entries of the journal come up uncolored (rwh)
* Fix 4861: Incremental find cycles through words while typing (uwog)

53

* Use 'instance' instead of 'data' as a path prefix; part of 4850 (uwog)

52

* Fix 4731: Find control should have embedded search/cancel icons (uwog)

51

* Fix 4614: Write should change the order of the zoom icons (uwog)
* Fix 4022: Write should not change the format of an opened journal entry (uwog)
* Remove tons of hacks to go with the new AbiWidget version (uwog)

50

* First implementation of 4165: Write must expose UI to 
  allow specifying output format (uwog)
* Update .pot file with new strings (uwog)

49

* Enable/disable the search functions based on the input (uwog)
* Support for searching text (foddex, tiny bit of uwog)
* Support custom keybindings (foddex)

48

* Fix 2124: Fail to page in Write (foddex)
* Fix 3899: do not use set_title (uwog)

47

* Advertise odt support (marco)

45

* Update spanish translation (beckerde)

44

* Update spanish translation (beckerde)

43

* Update spanish translation (beckerde)

42

* Initial german translation (fab)

41

* Fix #3319: default tab should be 'text' (uwog)

40

* Add first 3000 words indexing (tomeu)

39

* The part needed in write for viewing the source of browse #3261 (erikos)
* Updated the translations for new strings (uwog)

38

* Updated activity icon (erikos)

37

* (Temporary) add styles used by loaded documents that are not available 
  by default to the style dropdown list (uwog)
* (Temporary) add fonts used by loaded documents that are not available 
  on the system to the font dropdown list (uwog)
* Set a tooltip on various menu items (#821) (uwog)
* Hook up the abiword canvas 'style-name' signal to the Format toolbar (uwog)
* Hook up the abiword canvas 'font-family' signal to the Text toolbar (uwog)
* Add a format-text-size icon before the text size selection dropdown (uwog)
* Don't forget to actually show the separators in the Text toolbar (uwog)
* Add a Style label before the Style combobox (uwog)
* Hook up the abiword canvas 'font-size' signal to the Text toolbar (uwog)

36

* Use a journal object picker when inserting images, instead of a normal
  GTK+ Open File dialog (uwog)
* Port to new tubes API (cassidy)
* Add supported mimetypes for abiword, msword, xhtml, html and rtf (uwog)
* Add a format toolbar (uwog)
* Implement basic style support (uwog)
* Make style names translatable (uwog)

35

* Update zoom icon names
* Adapt icon names to the new API

34

* Add French translation. (Samuel Bizien)

33

* Add Greek translation. (simosx)

32

* #2370 Add spanish translation (xavi)

31

* Add macedonian translation.

30

* Update some icons (uwog)
* Implement full justification (uwog)
* Workaround for missing buddy removed in _on_dbus_names_changed (uwog)
* Implement buddies leaving (uwog)
* Workaround for wrong view margin size calculation (uwog)
* Remove the border around the view (uwog)
* Connect to the zoom signal to show the currect zoom level (uwog)

29

* Use a logger for the toolbar (uwog)

28

* add total page count label (uwog)
* replace print's with proper logger calls (uwog)
* Implement a color button, which respects the cursors context (uwog)

27

* Make sure the table icon shows up (uwog)
* Use ToolComboBoxes for ComboBoxes om the toolbar (uwog)

25

* Add brazilian translation. (DiegoZacarao)

24

* Set the right mime type when saving to the journal. (tomeu)

23

* Arabian translation.
* Adapt to combobox API change.


widgets.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import os
import dbus
import time
from gettext import gettext as _
import logging

import gi
try:
    gi.require_version('Abi', '2.9')
except:
    gi.require_version('Abi', '3.0')
from gi.repository import Abi
from gi.repository import GLib

from sugar3.graphics.radiotoolbutton import RadioToolButton
from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.palettemenu import PaletteMenuItem
from sugar3.datastore import datastore

from sugar3.activity.activity import SCOPE_PRIVATE

logger = logging.getLogger('write-activity')


class AbiButton(RadioToolButton):

    def __init__(self, abi, abi_signal, do_abi_cb, on_abi_cb=None, **kwargs):
        RadioToolButton.__init__(self, **kwargs)

        self._abi_handler = abi.connect(abi_signal, self.__abi_cb,
                                        abi_signal, on_abi_cb)
        self._toggled_handler = self.connect('toggled', self.__toggled_cb,
                                             abi, do_abi_cb)

    def __toggled_cb(self, button, abi, do_abi_cb):
        if not button.props.active:
            return

        abi.handler_block(self._abi_handler)
        try:
            logging.debug('Do abi %s' % do_abi_cb)
            do_abi_cb()
        finally:
            abi.handler_unblock(self._abi_handler)

    def __abi_cb(self, abi, prop, abi_signal, on_abi_cb):
        if (on_abi_cb is None and not prop) or \
                (on_abi_cb is not None and not on_abi_cb(abi, prop)):
            return

        self.handler_block(self._toggled_handler)
        try:
            logging.debug('On abi %s prop=%r' % (abi_signal, prop))
            self.set_active(True)
        finally:
            self.handler_unblock(self._toggled_handler)


class AbiMenuItem(PaletteMenuItem):

    def __init__(self, abi, abi_signal, do_abi_cb, icon_name, label,
                 button, on_abi_cb=None, button_icon_name=None):
        self._icon_name = icon_name
        # _button_icon_name is used only in the first case of
        # the list menu
        self._button_icon_name = button_icon_name
        self._button = button
        PaletteMenuItem.__init__(self, icon_name=icon_name, text_label=label)

        self._abi_handler = abi.connect(abi_signal, self.__abi_cb,
                                        abi_signal, on_abi_cb)
        self.connect('activate', self.__activated_cb, abi, do_abi_cb)

    def __activated_cb(self, button, abi, do_abi_cb):

        if self._button_icon_name is not None:
            if self._button.get_icon_name() == self._button_icon_name:
                return
        else:
            if self._button.get_icon_name() == self._icon_name:
                return

        abi.handler_block(self._abi_handler)
        try:
            logging.debug('Do abi %s' % do_abi_cb)
            do_abi_cb()
            if self._button_icon_name is not None:
                self._button.set_icon_name(self._button_icon_name)
            else:
                self._button.set_icon_name(self._icon_name)
        finally:
            abi.handler_unblock(self._abi_handler)

    def __abi_cb(self, abi, prop, abi_signal, on_abi_cb):
        if (on_abi_cb is None and not prop) or \
                (on_abi_cb is not None and not on_abi_cb(abi, prop)):
            return

        logging.debug('On abi %s prop=%r' % (abi_signal, prop))
        if self._button_icon_name is not None:
            self._button.set_icon_name(self._button_icon_name)
        else:
            self._button.set_icon_name(self._icon_name)


class ExportButtonFactory():

    _EXPORT_FORMATS = [{'mime_type': 'application/rtf',
                        'title': _('Rich Text (RTF)'),
                        'icon': 'save-as-rtf',
                        'jpostfix': _('RTF'),
                        'exp_props': ''},

                       {'mime_type': 'text/html',
                        'title': _('Hypertext (HTML)'),
                        'icon': 'save-as-html',
                        'jpostfix': _('HTML'),
                        'exp_props': 'html4:yes; declare-xml:no; '
                                     'embed-css:yes; embed-images:yes;'},

                       {'mime_type': 'text/plain',
                        'title': _('Plain Text (TXT)'),
                        'icon': 'save-as-txt',
                        'jpostfix': _('TXT'),
                        'exp_props': ''},

                       {'mime_type': 'application/pdf',
                        'title': _('Portable Document Format (PDF)'),
                        'icon': 'save-as-pdf',
                        'jpostfix': _('PDF'),
                        'exp_props': ''}]

    def __init__(self, activity, abi):

        toolbar = activity.activity_button.props.page
        for i in self._EXPORT_FORMATS:
            if abi.get_version() == '3.0' and i['title'].find('PDF') > -1:
                # pdf export crashes on abiword 3.0
                continue
            button = ToolButton(i['icon'])
            button.set_tooltip(i['title'])
            button.connect('clicked', self.__clicked_cb, activity, abi, i)
            toolbar.insert(button, -1)
            button.show()

    def __clicked_cb(self, menu_item, activity, abi, format):
        logger.debug('exporting file: %r' % format)

        exp_props = format['exp_props']

        # special case HTML export to set the activity name as the HTML title
        if format['mime_type'] == "text/html":
            exp_props += " title:" + activity.metadata['title'] + ';'

        # create a new journal item
        fileObject = datastore.create()
        act_meta = activity.metadata
        fileObject.metadata['title'] = \
            act_meta['title'] + ' (' + format['jpostfix'] + ')'
        fileObject.metadata['title_set_by_user'] = \
            act_meta['title_set_by_user']
        fileObject.metadata['mime_type'] = format['mime_type']
        fileObject.metadata['fulltext'] = \
            abi.get_content('text/plain', None)[0][:3000]

        fileObject.metadata['icon-color'] = act_meta['icon-color']

        # don't set application if PDF because Write can't open PDF files
        if format['mime_type'] != 'application/pdf':
            fileObject.metadata['activity'] = act_meta['activity']

        fileObject.metadata['keep'] = act_meta.get('keep', '0')

        preview = activity.get_preview()
        if preview is not None:
            fileObject.metadata['preview'] = dbus.ByteArray(preview)

        fileObject.metadata['share-scope'] = act_meta.get('share-scope',
                                                          SCOPE_PRIVATE)

        # write out the document contents in the requested format
        fileObject.file_path = os.path.join(activity.get_activity_root(),
                                            'instance', '%i' % time.time())
        abi.save('file://' + fileObject.file_path,
                 format['mime_type'], exp_props)

        # store the journal item
        datastore.write(fileObject, transfer_ownership=True)
        fileObject.destroy()
        del fileObject


class DocumentView(Abi.Widget):

    def __init__(self):
        Abi.init([])
        Abi.Widget.__init__(self)
        self.connect('size-allocate', self.__size_allocate_cb)
        try:
            self.connect('request-clear-area', self.__request_clear_area_cb)
        except:
            logging.debug('EXCEPTION: request-clear-area signal not available')

        try:
            self.connect('unset-clear-area', self.__unset_clear_area_cb)
        except:
            logging.debug('EXCEPTION: unset-clear-area signal not available')

        self.osk_changed = False
        self.dy = 0

    def __shallow_move_cb(self):
        self.moveto_right()
        return False

    def __size_allocate_cb(self, widget, allocation):
        self.set_allocation(allocation)

        if self.get_child() is not None:
            child_allocation = allocation
            child_allocation.y = 0
            child_allocation.x = 0
            child_allocation.height -= self.dy
            self.get_child().size_allocate(allocation)

        if self.osk_changed is True:
            self.moveto_left()
            GLib.timeout_add(100, self.__shallow_move_cb)
            self.osk_changed = False

    def __request_clear_area_cb(self, widget, clear, cursor):
        allocation = widget.get_allocation()
        allocation.x = 0
        allocation.y = 0
        allocation.x, allocation.y = \
            widget.get_window().get_root_coords(allocation.x, allocation.y)

        if clear.y > allocation.y + allocation.height or \
                clear.y + clear.height < allocation.y:
            return False

        self.dy = allocation.y + allocation.height - clear.y

        # Ensure there's at least some room for the view
        if self.dy > allocation.height - 80:
            self.dy = 0
            return False

        self.osk_changed = True
        self.queue_resize()
        return True

    def __unset_clear_area_cb(self, widget, snap_back):
        self.dy = 0
        self.queue_resize()
        return True

    def get_version(self):
        version = Abi._version
        logging.debug('Abiword version %s', version)
        return version
